app/javascript/packages/memorable-date/index.ts

```diff
@@ -0,0 +1,386 @@
+  }
+
+  /**
+   * HTML input element for entering the day part of the date
+   */
+  get dayInput(): HTMLInputElement | null {
+    return this.querySelector('.memorable-date__day');
+  }
+
+  /**
+   * HTML input element for entering the year part of the date
+   */
+  get yearInput(): HTMLInputElement | null {
+    return this.querySelector('.memorable-date__year');
+  }
+
+  /**
+   * List of HTML input elements for entering month, day and year for a date
+   */
+  get allInputs(): HTMLInputElement[] {
+    const month = this.monthInput;
+    const day = this.dayInput;
+    const year = this.yearInput;
+    const inputs: HTMLInputElement[] = [];
+
+    month && inputs.push(month);
+    day && inputs.push(day);
+    year && inputs.push(year);
+
+    return inputs;
+  }
+
+  /**
+   * The configured minimum valid value for the date, based on "min" HTML attribute
+   */
+  get min(): Date | null {
+    return this.getDateAttribute('min');
+  }
+
+  /**
+   * The configured maximum valid value for the date, based on "max" HTML attribute
+   */
+  get max(): Date | null {
+    return this.getDateAttribute('max');
+  }
+
+  connectedCallback() {
+    // Wait to show validation until first submission
+    let updateErrorMessage = false;
+
+    const { allInputs } = this;
+
+    const inputListener = (e: Event) => {
+      // Run validations
+      this.validate();
+
+      // Update error messages on form validation
+      if (e.type === 'invalid') {
+        updateErrorMessage = true;
+      }
+
+      if (updateErrorMessage) {
+        // Trigger error message updates
+        const errorMessage = allInputs.find((f) => f.validationMessage)?.validationMessage;
+        allInputs.forEach((f) => {
+          f.closest('lg-validated-field')?.setErrorMessage(errorMessage || null);
+        });
+      }
+    };
+
+    allInputs.forEach((i) => {
+      i.addEventListener('input', inputListener);
+      i.addEventListener('invalid', inputListener);
+    });
+  }
+
+  validate(): void {
+    const month = this.monthInput;
+    const day = this.dayInput;
+    const year = this.yearInput;
+
+    if (!(month && day && year)) {
+      // Cannot accurately run validation w/o all fields
+      return;
+    }
+
+    const { error_messages: errorMessages, range_errors: rangeErrors } =
+      this.getErrorMessageMappings();
+
+    const hasMissingValues = [
+      { month, day, year },
+      { month, day },
+      { month, year },
+      { month },
+      { day, year },
+      { day },
+      { year },
+    ].some(this.checkMissingValues(errorMessages));
+
+    if (hasMissingValues) {
+      return;
+    }
+
+    const hasInvalidValues = [{ month }, { day }, { year }].some(
+      this.checkFieldsInvalid(errorMessages),
+    );
+
+    if (hasInvalidValues) {
+      return;
+    }
+
+    let parsedDate: Date | undefined;
+    try {
+      const parsedUnixTime = Date.parse(`${year.value}-${month.value}-${day.value}`);
+      parsedDate = new Date(parsedUnixTime);
```

Comment:
I think we could skip the `Date.parse` and construct the date directly with the string value, since [the constructor supports a value which would be parseable by `Date.parse`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/Date#datestring).

```suggestion
      parsedDate = new Date(`${year.value}-${month.value}-${day.value}`);
```

Reply:
Fair point. I had learned to distrust the `Date` constructor due to browser differences, but looks like we can do this b/c it's in the ECMAScript spec.

Reply:
Updated to use the `Date` constructor instead of `Date.parse`.